prefect_gcp/cloud_storage.py

```diff
@@ -452,14 +451,195 @@ def example_cloud_storage_copy_blob_flow():
 
     source_blob_obj = source_bucket_obj.blob(source_blob)
-    partial_copy_blob = partial(
+    await run_sync_in_worker_thread(
         source_bucket_obj.copy_blob,
         blob=source_blob_obj,
         destination_bucket=dest_bucket_obj,
         new_name=dest_blob,
         timeout=timeout,
         **copy_kwargs,
     )
-    await to_thread.run_sync(partial_copy_blob)
 
     return dest_blob
+
+
+class GcsBucket(ReadableFileSystem, WritableFileSystem):
+    """
+    Block used to store data using GCP Cloud Storage Buckets.
+
+    Attributes:
+        bucket: Name of the bucket.
+        gcp_credentials: The credentials to authenticate with GCP.
+        basepath: Used when you don't want to read/write at base level.
+
+    Example:
+        Load stored GCP Cloud Storage Bucket:
+        ```python
+        from prefect_gcp import GcpCloudStorageBucket
+        gcp_cloud_storage_bucket_block = GcpCloudStorageBucket.load("BLOCK_NAME")
+        ```
+    """
+
+    _logo_url = "https://images.ctfassets.net/gm98wzqotmnx/4CD4wwbiIKPkZDt4U3TEuW/c112fe85653da054b6d5334ef662bec4/gcp.png?h=250"  # noqa
+    _block_type_name = "GCS Bucket"
+
+    bucket: str
+    gcp_credentials: GcpCredentials
+    basepath: str = ""
+
+    def _resolve_path(self, path: str) -> str:
+        """
+        A helper function used in write_path to join `self.basepath` and `path`.
+
+        Args:
+            path: Name of the key, e.g. "file1". Each object in your
+                bucket has a unique key (or key name).
+
+        Returns:
+            The joined path.
+        """
+        path = path or str(uuid4())
+
+        # If basepath provided, it means we won't write to the root dir of
+        # the bucket. So we need to add it on the front of the path.
+        path = os.path.join(self.basepath, path) if self.basepath else path
+        return path
+
+    @sync_compatible
+    async def get_directory(
+        self, from_path: Optional[str] = None, local_path: Optional[str] = None
+    ) -> None:
+        """
+        Copies a folder from the configured GCS bucket to a local directory.
+        Defaults to copying the entire contents of the block's basepath to the current
+        working directory.
+
+        Args:
+            from_path: Path in GCS bucket to download from. Defaults to the block's
+                configured basepath.
+            local_path: Local path to download GCS bucket contents to.
+                Defaults to the current working directory.
+        """
+        from_path = (
+            self.basepath if from_path is None else self._resolve_path(from_path)
+        )
+
+        if local_path is None:
+            local_path = os.path.abspath(".")
+        else:
+            local_path = os.path.expanduser(local_path)
+
+        project = self.gcp_credentials.project
+        client = self.gcp_credentials.get_cloud_storage_client(project=project)
+
+        for blob in client.list_blobs(self.bucket):
```

Comment:
Might want to run this in a worker thread since I suspect it will be blocking.

Also, should `from_path` be used here?